File: templates/request.php
<?php
/*
 *  Free Room Finder
 *
 *  Copyright (C) 2013 Jonathan Gillett and Joseph Heron
 *  All rights reserved.
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Affero General Public License as
 *  published by the Free Software Foundation, either version 3 of the
 *  License, or (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Affero General Public License for more details.
 *
 *  You should have received a copy of the GNU Affero General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */


/*
 * The room request/search form, used to search for an available room on campus
 * 
 * 
 * DEPENDENCIES
 * ------------
 * 
 * This template depends on the array $campus being set with the list of available campuses
 * 
 * $campuses = array(	'UOIT North Campus',
 *						'UOIT Downtown Campus',
 *						...
 *					);
 * 
 */

/* The following are the available times, durations, and number of people */
/* TODO FIX THIS UGLY HACK */
$avail_times = array(	"08:00:00",
						"09:00:00",
						"10:00:00",
						"11:00:00",
						"12:00:00",
						"13:00:00",
						"14:00:00",
						"15:00:00",
						"16:00:00",
						"17:00:00",
						"18:00:00",
						"19:00:00",
						"20:00:00",
						"21:00:00",
						"22:00:00"
					);

$durations = array (1, 2, 3);
$num_people = array (1, 2, 3, 4, 5);
?>
<section id="request_room">
	<div class="page-header">
		<h1>Request Room</h1>
	</div>
	<div class="row">
		<div class="span8">
			<form class="well form-horizontal" action="room_results.php" method="post" accept-charset="UTF-8">
				<fieldset>
					<div class="control-group">
			      		<label for="select_time" class="control-label">Time</label>
				      	<div class="controls">
				      		<select id="select_time" name="select_time" class="input-xlarge">
				      			<?php
				      				/* Set the default option as the current time */
					      			date_default_timezone_set('America/Toronto');
					      			$cur_time = date('H:i:s');
					      			echo '<option>' . $cur_time . '</option>';
				      				foreach ($avail_times as $time)
				      				{
				      					echo '<option>' . $time . '</option>';
				      				}
				      			?>
				      		</select>
				      	</div>
		      		</div>
		      		<div class="control-group">
			      		<label for="select_duration" class="control-label">Duration</label>
				      	<div class="controls">
				      		<div class="input-append">
					      		<select id="select_duration" name="select_duration" class="input-xlarge">
					      			<?php
					      				foreach ($durations as $duration)
					      				{
					      					echo '<option>' . $duration . '</option>';
					      				}
					      			?>
					      		</select>
					      		<span class="add-on">Hours</span>
				      		</div>
				      	</div>
		      		</div>
		      		<div class="control-group">
			      		<label for="select_date" class="control-label">Date</label>
							<div class="controls">
							<class id="dp3" data-date-format="yyyy-mm-dd">
							<!--  Set the current date -->
							<?php
								date_default_timezone_set('America/Toronto');
								$cur_date = date('Y-m-d');
								echo '<input id="select_date" name="select_date"class="input-datelarge" type="text" value="'. $cur_date .'">';
 							?>
							</div>
		      		</div>
		      		<div class="control-group">
			      		<label for="select_campus" class="control-label">Campus</label>
				      	<div class="controls">
				      		<select id="select_campus" name="select_campus" class="input-xlarge">
				      			<?php
				      				foreach ($campuses as $campus)
				      				{
				      					echo '<option>' . $campus . '</option>';
				      				}
				      			?>
				      		</select>
				      	</div>
		      		</div>
		      		<div class="control-group">
			      		<label for="select_num_people" class="control-label">Number of People</label>
				      	<div class="controls">
				      		<select id="select_num_people" name="select_num_people" class="input-xlarge">
				      			<?php
				      				foreach ($num_people as $num)
				      				{
				      					echo '<option>' . $num . '</option>';
				      				}
				      			?>
				      		</select>
				      	</div>
		      		</div>

		      		<div class="form-actions">
		            	<button class="btn btn-primary" type="submit" name="search_room" value="Search">Search</button>
		          </div>
				</fieldset>
			</form>
		</div>
	</div>
</section>

File: MobileApp/src/com/uoit/freeroomfinder/Request.java
/**
 * Free Room Finder (FRF)
 * Tired of rooms on campus always being in use? Fear no more the FRF is here.
 *
 * Copyright (C) 2013 Joseph Heron, Jonathan Gillett, and Daniel Smullen
 * All rights reserved.
 *
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.uoit.freeroomfinder;

import java.util.ArrayList;
import java.util.List;
import retrofit.RestAdapter;
import retrofit.http.GET;
import retrofit.http.Path;

/**
 * Request An abstract class for formulating requests to the external web data.
 * 
 * @author Joseph Heron
 * @author Jonathan Gillett
 * @author Daniel Smullen
 * 
 */
public abstract class Request
{
    /**
     * The URL for the back-end API.
     */
    private static final String API_URL = "http://cs-club.ca/free-room-website/api";

    /**
     * Credentials A static class representing the JSON encoded login credentials.
     */
    private static class Credentials
    {
        /**
         * Stores whether the credentials are valid or not.
         */
        boolean credentials = false;
    }

    /**
     * Room A static class representing the JSON encoded rooms.
     */
    private static class Room
    {
        /**
         * The room name and number.
         */
        String room;
        
        /**
         * The availability start time.
         */
        String starttime;
        
        /**
         * The availability end time.
         */
        String endtime;
    }

    /**
     * LoginCredentials A REST API interface for getting login credentials
     */
    private static interface LoginCredentials
    {
        // Provide the formatted login credentials query string.
        @GET("/login/{username}/{password}")
        Credentials loginCredentials(@Path("username") String username,
                @Path("password") String password);
    }

    /**
     * AvailableRooms A REST API interface for searching for rooms
     */
    private static interface AvailableRooms
    {
        // Provide the formatted query string for the available rooms.
        @GET("/availablerooms/{time}/{date}/{campus}/{duration}")
        List<Room> availableRooms(@Path("time") String time, @Path("date") String date,
                @Path("campus") String campus, @Path("duration") int duration);
    }

    /**
     * validateCredentials Validates the login credentials of the user.
     * 
     * @param username The user's user name.
     * @param password The user's password.
     * @return Returns true if the credentials are valid, false otherwise.
     */
    public static boolean validateCredentials(String username, String password)
    {
        Credentials credentials = new Credentials();
        RestAdapter restAdapter = new RestAdapter.Builder().setServer(API_URL).build();

        try
        {
            /**
             * Create an instance of the REST API interface
             */
            LoginCredentials loginCredentials = restAdapter.create(LoginCredentials.class);

            /**
             * Validate the credentials
             */
            credentials = loginCredentials.loginCredentials(username, password);
        }
        catch (Exception e)
        {
            e.printStackTrace();
        }

        return credentials.credentials;
    }

    /**
     * searchRooms A simplified interface to search for the rooms using the REST API.
     * 
     * @param time The specified time that a room is needed.
     * @param date The specified date that a room is needed.
     * @param campus The campus the room is located in.
     * @param duration The length of time the room is needed for.
     * 
     * @return Returns the list of available rooms that match the query.
     */
    public static ArrayList<Rooms> searchRooms(String time, String date, String campus, int duration)
    {
        ArrayList<Rooms> results = new ArrayList<Rooms>();

        /* Create a REST adapter which points the Free Room finder API */
        RestAdapter restAdapter = new RestAdapter.Builder().setServer(API_URL).build();

        try
        {
            /* Create an instance of the REST API interface */
            AvailableRooms availableRooms = restAdapter.create(AvailableRooms.class);

            /* Search for available rooms */
            List<Room> rooms = availableRooms.availableRooms(time, date, campus, duration);
            for (Room room : rooms)
            {
                results.add(new Rooms(room.room, DateTimeUtility.parseFullTime(room.starttime)
                        .getTime(), DateTimeUtility.parseFullTime(room.endtime).getTime(),
                        DateTimeUtility.parseDate(date).getTime()));
            }
        }
        catch (Exception e)
        {
            e.printStackTrace();
        }

        return results;
    }
}
